Tests/DiligentToolsTest/src/TestingEnvironment.cpp

```diff
@@ -0,0 +1,105 @@
+ *  contract, or otherwise, unless required by applicable law (such as deliberate
+ *  and grossly negligent acts) or agreed to in writing, shall any Contributor be
+ *  liable for any damages, including any direct, indirect, special, incidental,
+ *  or consequential damages of any character arising as a result of this License or
+ *  out of the use or inability to use the software (including but not limited to damages
+ *  for loss of goodwill, work stoppage, computer failure or malfunction, or any and
+ *  all other commercial damages or losses), even if such Contributor has been advised
+ *  of the possibility of such damages.
+ */
+
+#include "TestingEnvironment.hpp"
+
+namespace Diligent
+{
+
+namespace Testing
+{
+
+TestingEnvironment::TestingEnvironment()
+{
+    SetDebugMessageCallback(MessageCallback);
+}
+
+TestingEnvironment::~TestingEnvironment()
+{
+    SetDebugMessageCallback(nullptr);
+}
+
+TestingEnvironment* TestingEnvironment::Initialize(int argc, char** argv)
+{
+    m_pTheEnvironment = new TestingEnvironment{};
+    return m_pTheEnvironment;
+}
+
+TestingEnvironment* TestingEnvironment::GetInstance() { return m_pTheEnvironment; }
+
+void TestingEnvironment::SetErrorAllowance(int NumErrorsToAllow, const char* InfoMessage)
+{
+    m_NumAllowedErrors = NumErrorsToAllow;
+    if (InfoMessage != nullptr)
+        std::cout << InfoMessage;
+    if (m_NumAllowedErrors == 0)
+    {
+        std::unique_lock<std::mutex> Lock{m_ExpectedErrorSubstringsMtx};
+        m_ExpectedErrorSubstrings.clear();
+    }
+}
+
+
+void TestingEnvironment::PushExpectedErrorSubstring(const char* Str, bool ClearStack)
+{
+    if (ClearStack)
+        m_ExpectedErrorSubstrings.clear();
+    VERIFY_EXPR(Str != nullptr && Str[0] != '\0');
+    m_ExpectedErrorSubstrings.push_back(Str);
+}
+
+void TestingEnvironment::MessageCallback(DEBUG_MESSAGE_SEVERITY Severity, const Char* Message, const char* Function, const char* File, int Line)
+{
+    if (Severity == DEBUG_MESSAGE_SEVERITY_ERROR || Severity == DEBUG_MESSAGE_SEVERITY_FATAL_ERROR)
+    {
+        if (m_NumAllowedErrors == 0)
+        {
+            ADD_FAILURE() << "Unexpected error";
+        }
+        else
+        {
+            m_NumAllowedErrors--;
+            std::unique_lock<std::mutex> Lock{m_ExpectedErrorSubstringsMtx};
+            if (!m_ExpectedErrorSubstrings.empty())
+            {
+                const auto& ErrorSubstring = m_ExpectedErrorSubstrings.back();
+                if (strstr(Message, ErrorSubstring.c_str()) == nullptr)
+                    ADD_FAILURE() << "Expected error substring '" << ErrorSubstring << "' was not found in the error message";
+                m_ExpectedErrorSubstrings.pop_back();
+            }
+        }
+    }
+
+    PlatformDebug::OutputDebugMessage(Severity, Message, Function, File, Line);
+}
+
+TestingEnvironment*      TestingEnvironment::m_pTheEnvironment;
+std::atomic_int          TestingEnvironment::m_NumAllowedErrors;
```

Comment:
Even though static members are initialized to zero, prefer to explicitly initialize them for consistency and clarity:
```cpp
TestingEnvironment*      TestingEnvironment::m_pTheEnvironment = nullptr;
std::atomic_int          TestingEnvironment::m_NumAllowedErrors{0};
```